File: netlogo-gui/src/main/prim/_repeat.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.core.Let;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableLong;

public final strictfp class _repeat
    extends Command
    implements org.nlogo.nvm.CustomAssembled {
  public final Let let = new Let(null);

  @Override
  public org.nlogo.core.Syntax syntax() {
    return Syntax.commandSyntax
        (new int[]{Syntax.NumberType(),
            Syntax.CommandBlockType()});
  }

  @Override
  public void perform(final Context context) throws LogoException {
    perform_1(context, argEvalDoubleValue(context, 0));
  }

  public void perform_1(final Context context, double d0) throws LogoException {
    context.let(let, new MutableLong(validLong(d0)));
    context.ip = offset;
  }

  @Override
  public String toString() {
    return super.toString() + ":+" + offset;
  }

  public void assemble(org.nlogo.nvm.AssemblerAssistant a) {
    a.add(this);
    a.block();
    a.resume();
    a.add(new _repeatinternal(1 - a.offset(), let));
  }
}

File: netlogo-gui/src/main/prim/_repeat.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.{Let, Token}
import org.nlogo.api.LogoException
import org.nlogo.nvm.{AssemblerAssistant, Command, Context, CustomAssembled, MutableLong}
import org.nlogo.prim.etc.{_enterscope, _exitscope}

class _repeat(_token: Token) extends Command with CustomAssembled {
  token_=(_token)

  private[this] val _let: Let = Let("~" + _token.text + "_" + _token.start.toString)

  def let = _let

  @throws(classOf[LogoException])
  override def perform(context: Context): Unit = {
    perform_1(context, argEvalDoubleValue(context, 0))
  }

  @throws(classOf[LogoException])
  def perform_1(context: Context, d0: Double): Unit = {
    context.activation.binding.let(_let, MutableLong(validLong(d0, context)))
    context.ip = offset
  }

  override def toString: String = {
    super.toString() + ":+" + offset
  }

  def assemble(a: AssemblerAssistant): Unit = {
    a.add(this)
    a.add(new _enterscope)
    a.block()
    a.add(new _exitscope)
    a.resume()
    a.add(new _repeatinternal(1 - a.offset, _let))
  }
}
